### PR TITLE
[PR] divramod/feat/importer (wip)

### PR DESCRIPTION
**Pull request by divramod**
solves #4

### CODE REVIEW

File: plugins/importer.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    return True
+
+def importer_import(iN):
+
+    _helper_log(f"IMPORT STARTED")
+    timeStartImport = dt.datetime.now()
+
+    # --- define vars
+    model = iN.get_model()
+    tablename = model.get_node("root").get_child('importer').get_child('tablename').get_value()
```
Comment (smartyal) [inline plugins/importer.py:58]: please don't use absolute path, instead:
``` tablename = model.get_node("root").get_child('importer').get_child('tablename').get_value() ```

=>

``` tableName = iN.get_parent().get_child("tablename") ```

File: plugins/importer.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    table = model.get_node("root").get_child('imports').get_child(tablename)
```
Comment (smartyal) [inline plugins/importer.py:59]: see above

File: plugins/importer.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    # --- create needed nodes
+    model.delete_node(f'root.imports.{tablename}.variables')
```
Comment (smartyal) [inline plugins/importer.py:62]: the "imports" folder should be inside the plug-in folder,

File: web/plugins/importer/index.js [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+    tablepath = 'root.' + $("#importer-tablepath").val().replace('/', '.') + tablename
+  }
+  const importerObj = { fields, filename, timefield, headerexists: importerHeaderExists }
+  // --- update html
+  msg = 'Waiting for import to be finished!'
+  $(selector).html(_helper_html_wrap(msg, btnHtml))
+  // --- create response var
+  let res 
+  // --- [api] create table node
+  res = http_post_sync('/_create', true, [ { browsePath: tablepath, type: 'table' } ])
```
Comment (smartyal) [inline web/plugins/importer/index.js:324]: create all objects/folder from the backend, place needed info in the metadata (or name it different like "importinfo"..)

File: web/plugins/importer/index.js [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+  if ( res.status > 201 ) $(selector).html(_helper_html_wrap(msg, btnHtml))
+  // --- [api] set fields
+  if (res.status <= 201)
+    res = http_post_sync('/setProperties', true, [ { browsePath: tablepath + ".metadata", value: JSON.stringify(importerObj) } ])
+  msg = `Failed setting value for const '${tablepath + ".metadata"}'!`
+  if ( res.status > 201 ) $(selector).html(_helper_html_wrap(msg, btnHtml))
+  // --- [api] run importer function
+  if (res.status <= 201) {
+      http_post( "/_execute", cockpitPath + ".importer_import", null, null, (self, status, data, params) => {
+        if ( status === 200 ) {
```
Comment (smartyal) [inline web/plugins/importer/index.js:350]: 200 only gives the succesfull start of the execution, please wait for an event by an observer, 
use the existing observer for this, make sure it has the right "target" set: importer_import..

File: web/plugins/importer/index.js [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+// --- ensures that the corresponding checkboxes have the right value
+// - when checkbox time is set, all other time checkboxes will be removed 
+// - the corresponding import checkbox is activated
+function _helper_checkbox_time(fieldId) {
+  const clickedElId = 'importer-field-time-' + fieldId
+  $("#importer-field-import-" + fieldId).attr('checked', 'checked')
+  $("input[id^=importer-field-time-]").each(function(index, el) {
+    const elId = $(el).attr('id')
+    if ( elId !== clickedElId ) {
+      $(el).removeAttr('checked')
```
Comment (smartyal) [inline web/plugins/importer/index.js:428]: auto-select of fields does not work when selecting the second time

File: plugins/importer.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    _helper_log(f"filename: {filename}")
+    _helper_log(f"timefield: {timefield}")
+    _helper_log(f"fields: {fields}")
+    _helper_log(f"data: {data}")
+    #  _helper_log(f"filepath: {filepath}")
+    #  _helper_log(f"dataFrame: {df}")
+    #  _helper_log(f"timeList: {timeList}")
+
+    for attr, value in data.items():
+        print(attr, value)
```
Comment (smartyal) [inline plugins/importer.py:111]: please no print, use logging
generally, use logging

File: plugins/importer.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    #  }
+
+    #  # list comprehension
+    #  epochs = [date2secs(time) for time in data["time"]]
+
+    #  # set values
+    #  val1.set_time_series(values=data["val1"],times=epochs)
+    #  val2.set_time_series(values=data["val2"],times=epochs)
+
+    time.sleep(1)
```
Comment (smartyal) [inline plugins/importer.py:127]: remove wait

File: plugins/importer.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    timefield = int(metadata["timefield"]) - 1
+    filename = metadata["filename"]
+    headerexists = metadata["headerexists"]
+    filepath = 'upload/' + filename 
+
+    # --- load csv data
+    # * https://www.shanelynn.ie/python-pandas-read_csv-load-data-from-csv-files/
+    # * [ ] optimize speed? https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas
+    # * [ ] vectorize a loop https://stackoverflow.com/questions/27575854/vectorizing-a-function-in-pandas
+    df = pd.read_csv(filepath)
```
Comment (smartyal) [inline plugins/importer.py:82]: please check if you get a progress callback, if so , use the "control.progress" node to write a value from 0...1

File: plugins/importer.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    #  # list comprehension
+    #  epochs = [date2secs(time) for time in data["time"]]
+
+    #  # set values
+    #  val1.set_time_series(values=data["val1"],times=epochs)
+    #  val2.set_time_series(values=data["val2"],times=epochs)
+
+    time.sleep(1)
+    _helper_log(f"IMPORT DONE (seconds: {(dt.datetime.now()-timeStartImport).seconds})")
+    return True
```
Comment (smartyal) [inline plugins/importer.py:129]: if you write a message to "control.result", it is displayed on the UI status line

File: web/js/tree.js [hunk 11]
```diff
@@ -1378,7 +1378,7 @@ class TreeCard
 
         //now start automatic updates
         this.autoUpdate = false; // we start without
-        //this.update_on_click(); //start the auto update
```
Comment (smartyal) [inline web/js/tree.js:1381]: please no auto start of the live update, redo the commenting

File: plugins/importer.py (hunk 9 above)
Comment (divramod) [reply plugins/importer.py:82]: i am not really sure if it is possible.
maybee this is something going into the right direction?
https://stackoverflow.com/questions/18603270/progress-indicator-during-pandas-operations

File: plugins/importer.py (hunk 10 above)
Comment (divramod) [reply plugins/importer.py:129]: can you point me to a code-example how you do that in other cases?

File: web/plugins/importer/index.js (hunk 5 above)
Comment (divramod) [reply web/plugins/importer/index.js:350]: i am not experienced with using the observer pattern here. i think we should do this one while pair-programming for saving time

File: web/plugins/importer/index.js (hunk 6 above)
Comment (divramod) [reply web/plugins/importer/index.js:428]: i wasn't able to reproduce it.

File: plugins/importer.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+def importer_preview(functionNode):
+
+    # --- set vars
+    fileName = 'upload/' + functionNode.get_child("fileName").get_value()
+    pathBase = os.getcwd()
+
+    # --- load csv data
+    dataFile = pd.read_csv(fileName)
+    previewData = dataFile.head()
```
Comment (smartyal) [inline plugins/importer.py:42]: please dont read the whole file 
or at least write only some rows (as many as needed for the frontenc)

File: plugins/importer.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+    #  # --- create needed nodes
+    importerNode.create_child('imports', type="folder")
+    importsNode = importerNode.get_child("imports")
+    try:
+        importsNode.delete_node(tablename)
+        importsNode.delete_node(f'{tablename}.variables')
+        importsNode.delete_node(f'{tablename}.columns')
+    except:
+        print("[importer] error while trying to delete node table")
```
Comment (smartyal) [inline plugins/importer.py:72]: please not print, user logger

File: plugins/importer.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    # --- define vars
+    importerNode = iN.get_parent()
+
+    # --- [vars] define
+    tablename = iN.get_child("tablename").get_value()
+    _helper_log(f"tablename: {tablename}")
+
+    #  # --- create needed nodes
+    importerNode.create_child('imports', type="folder")
+    importsNode = importerNode.get_child("imports")
```
Comment (smartyal) [inline plugins/importer.py:66]: delete_node is part of the Model() class api, use nodexx.delete() instred:

tableNode = importerNode.get_child("imports").get_child(tablename)
if tableNode:  
  tableNode.delete()
..
continue creating stuff

File: plugins/importer.py (hunk 9 above)
Comment (divramod) [reply plugins/importer.py:97]:
```
functionNode.get_child("control").get_child("result").set_value("no grid information for this annotation")
```

Comment (divramod) [reply plugins/importer.py:97]: wenn er einmal angewaehlt gewesen ist

**Pull request merged**